### PR TITLE
2679 Add IntensityStats transform to record intensity statistics into meta data

### LINKED ISSUE #2679: Add IntensityStats transform
**Is your feature request related to a problem? Please describe.**
Currently, we have `original_shape`, `original_affine`, etc. in the meta_dict, I suggest to also store the `max`, `mean`, `std`, etc. of the intensity of input image in the meta dict, then the following transforms can **dynamically** use it as parameters to scale/shift/clip values instead of fixed value for all the images.
For example:
```py
[
    LoadImaged("image")
    EnsureChannelFirstd("image"),
    IntensityStatsd("image", key_prefix="orig", ops=["max", "std"]),
    RandScaleIntensityd("image"),
    RandShiftIntensityd("image", offset=0.1, src="orig_std"),
    ThresholdIntensityd("image", threshold="orig_max", above=True, cval="orig_max"),
]
```

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #2679 .

### Description
This PR added `IntensityStats` transform to record intensity statistics into meta data for other following transforms to dynamically adjust parameters for every image.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by Nic-Ma** [comment]
I enhanced `RandShiftIntensityd` transform to leverage the intensity stats at runtime, we can also enhance other intensity transforms later.

Thanks.

File: monai/transforms/intensity/array.py [hunk 1]
```diff
@@ -1618,3 +1627,58 @@ def __call__(self, img: np.ndarray):
...
+
+    def __call__(self, img: np.ndarray, meta_data: Optional[Dict] = None) -> Tuple[np.ndarray, Dict]:
+        if meta_data is None:
+            meta_data = {}
+
+        def _compute(op: Callable, img: np.ndarray):
+            if self.channel_wise:
+                return [op(c) for c in img]
+            else:
+                return op(img)
```
Comment (wyli) [inline monai/transforms/intensity/array.py:1674]: Do we need an option to compute from a masked area?

**Review by wyli — COMMENTED** [review]
Thanks, I put some comments inline

File: monai/transforms/intensity/array.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/transforms/intensity/array.py:1674]: Hi @wyli ,

Thanks for the suggestion, could you please help provide a simple example for this mask idea?

Comment (wyli) [reply monai/transforms/intensity/array.py:1674]: Sometimes we want to compute mean and std from only the foreground based on a segmentation mask or an intensity threshold, could this be achieved by a transform chain?

File: monai/transforms/intensity/array.py [hunk 2]
```diff
@@ -1618,3 +1627,58 @@ def __call__(self, img: np.ndarray):
                 img[h] = self.fill_value
 
         return img
+
+
+class IntensityStats(Transform):
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/transforms/intensity/array.py]: OK, changed to utility transform in the latest commit.
Thanks.

File: monai/transforms/intensity/array.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/transforms/intensity/array.py]: OK, let me add an option to apply a mask, about the threshold, it can be achieved by passing a callable function to `ops`.

Thanks.

File: monai/transforms/intensity/array.py [hunk 3]
```diff
@@ -1618,3 +1627,58 @@ def __call__(self, img: np.ndarray):
...
+        self.supported_ops = {
+            "mean": lambda x: np.nanmean(x),
+            "median": lambda x: np.nanmedian(x),
+            "max": lambda x: np.nanmax(x),
+            "min": lambda x: np.nanmin(x),
+            "std": lambda x: np.nanstd(x),
+        }
+        self.ops = ensure_tuple(ops)
+        for o in self.ops:
+            if isinstance(o, str) and o not in self.supported_ops:
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/transforms/intensity/array.py]: Updated in the latest commit.
Thanks.

File: monai/transforms/intensity/array.py [hunk 4]
```diff
@@ -1618,3 +1627,58 @@ def __call__(self, img: np.ndarray):
...
+            meta data dictionary. if some `ops` are callable functions, will use "{key_prefix}_custom_{index}"
+            as the key, where index counts from 0.
+        channel_wise: whether to compute statistics for every channel of input image separately.
+            if True, return a list of values for every operation, default to False.
+
+    """
+
+    def __init__(self, ops: Sequence[Union[str, Callable]], key_prefix: str, channel_wise: bool = False) -> None:
+        self.supported_ops = {
+            "mean": lambda x: np.nanmean(x),
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/transforms/intensity/array.py]: I added multi-processing test in the latest commit.
Seems it works fine.

Thanks.

File: monai/transforms/intensity/array.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/transforms/intensity/array.py]: Added `mask` option in the latest commit.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Thanks for your review.
I updated the PR according to your comments.
Could you please help review it again?

Thanks in advance.

File: tests/test_intensity_statsd.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            self.assertTrue(k in meta)
+            np.testing.assert_allclose(v, meta[k], atol=1e-3)
+
+    def test_dataloader(self):
+        dataset = Dataset(
+            data=[{"img": np.array([[[0.0, 1.0], [2.0, 3.0]]])}, {"img": np.array([[[0.0, 1.0], [2.0, 3.0]]])}],
+            transform=IntensityStatsd(keys="img", ops=["max", "mean"], key_prefix="orig"),
+        )
+        # set num workers = 0 for mac / win
+        num_workers = 2 if sys.platform == "linux" else 0
```
Comment (wyli) [inline tests/test_intensity_statsd.py:64]: I wonder if this test is enough to prevent us from this https://github.com/Project-MONAI/MONAI/issues/2122. I couldn't replicate the ticket 2122 even if I manually add lambda functions in the transform in line 61

Comment (Nic-Ma) [reply tests/test_intensity_statsd.py:64]: Hi @wyli ,

Thanks for pointing it, yes, I can reproduce the pickle issue with below program:
```py
from monai.transforms import (
    AsDiscrete,
    AddChanneld,
    Compose,
    CropForegroundd,
    IntensityStatsd,
    LoadImaged,
    Orientationd,
    RandCropByPosNegLabeld,
    ScaleIntensityRanged,
    Spacingd,
    ToTensord,
)
from monai.networks.nets import UNet
from monai.networks.layers import Norm
from monai.losses import DiceLoss
from monai.data import CacheDataset, DataLoader
from monai.apps import download_and_extract
import torch
import tempfile
import os
import glob
import torch.multiprocessing as mp


if __name__ == "__main__":

    root_dir = "/workspace/data/medical/"
    print(root_dir)


    resource = "https://msd-for-monai.s3-us-west-2.amazonaws.com/Task09_Spleen.tar"
    md5 = "410d4a301da4e5b2f6f86ec3ddba524e"


    compressed_file = os.path.join(root_dir, "Task09_Spleen.tar")
    data_dir = os.path.join(root_dir, "Task09_Spleen")
    if not os.path.exists(data_dir):
        download_and_extract(resource, compressed_file, root_dir, md5)

    train_images = sorted(glob.glob(os.path.join(data_dir, "imagesTr", "*.nii.gz")))
    train_labels = sorted(glob.glob(os.path.join(data_dir, "labelsTr", "*.nii.gz")))
    data_dicts = [
        {"image": image_name, "label": label_name}
        for image_name, label_name in zip(train_images, train_labels)
    ]
    train_files, val_files = data_dicts[:-9], data_dicts[-9:]


    train_transforms = Compose(
        [
            LoadImaged(keys=["image", "label"]),
            AddChanneld(keys=["image", "label"]),
            IntensityStatsd(keys="image", ops=["max", "min"], key_prefix="orig"),
            Spacingd(keys=["image", "label"], pixdim=(
                1.5, 1.5, 2.0), mode=("bilinear", "nearest")),
            Orientationd(keys=["image", "label"], axcodes="RAS"),
            ScaleIntensityRanged(
                keys=["image"], a_min=-57, a_max=164,
                b_min=0.0, b_max=1.0, clip=True,
            ),
            CropForegroundd(keys=["image", "label"], source_key="image"),
            RandCropByPosNegLabeld(
                keys=["image", "label"],
                label_key="label",
                spatial_size=(96, 96, 96),
                pos=1,
                neg=1,
                num_samples=4,
                image_key="image",
                image_threshold=0,
            ),
            ToTensord(keys=["image", "label"]),
        ]
    )
    val_transforms = Compose(
        [
            LoadImaged(keys=["image", "label"]),
            AddChanneld(keys=["image", "label"]),
            Spacingd(keys=["image", "label"], pixdim=(
                1.5, 1.5, 2.0), mode=("bilinear", "nearest")),
            Orientationd(keys=["image", "label"], axcodes="RAS"),
            ScaleIntensityRanged(
                keys=["image"], a_min=-57, a_max=164,
                b_min=0.0, b_max=1.0, clip=True,
            ),
            CropForegroundd(keys=["image", "label"], source_key="image"),
            ToTensord(keys=["image", "label"]),
        ]
    )

    train_ds = CacheDataset(
        data=train_files, transform=train_transforms,
        cache_rate=1.0, num_workers=4)
    # train_ds = monai.data.Dataset(data=train_files, transform=train_transforms)

    # use batch_size=2 to load images and use RandCropByPosNegLabeld
    # to generate 2 x 4 images for network training
    train_loader = DataLoader(train_ds, batch_size=2, shuffle=True, num_workers=4)

    val_ds = CacheDataset(data=val_files, transform=val_transforms, cache_rate=1.0, num_workers=4)
    val_loader = DataLoader(val_ds, batch_size=1, num_workers=4)

    device = torch.device("cuda:0")
    model = UNet(
        dimensions=3,
        in_channels=1,
        out_channels=2,
        channels=(16, 32, 64, 128, 256),
        strides=(2, 2, 2, 2),
        num_res_units=2,
        norm=Norm.BATCH,
    ).to(device)
    loss_function = DiceLoss(to_onehot_y=True, softmax=True)
    optimizer = torch.optim.Adam(model.parameters(), 1e-4)
    model.share_memory()


    max_epochs = 600
    val_interval = 2
    best_metric = -1
    best_metric_epoch = -1
    epoch_loss_values = []
    metric_values = []
    post_pred = AsDiscrete(argmax=True, to_onehot=True, n_classes=2)
    post_label = AsDiscrete(to_onehot=True, n_classes=2)

    for epoch in range(max_epochs):
        print("-" * 10)
        print(f"epoch {epoch + 1}/{max_epochs}")

        # use PyTorch multiprocessing thread instead
        try:
            mp.set_start_method('spawn', force=True)
        except ValueError:
            pass


        p = mp.Process(target=model.train, args=())
        p.start()
        p.join()

        epoch_loss = 0
        step = 0
        for batch_data in train_loader:
            step += 1
            inputs, labels = (
                batch_data["image"].to(device),
                batch_data["label"].to(device),
            )
            optimizer.zero_grad()
            outputs = model(inputs)
            loss = loss_function(outputs, labels)
            loss.backward()
            optimizer.step()
            epoch_loss += loss.item()
            print(
                f"{step}/{len(train_ds) // train_loader.batch_size}, "
                f"train_loss: {loss.item():.4f}")
        epoch_loss /= step
        epoch_loss_values.append(epoch_loss)
        print(f"epoch {epoch + 1} average loss: {epoch_loss:.4f}")
```
So I updated the PR to fix it in the latest commit, now it runs fine.
Could you please help review it again?

Thanks.

**Review by wyli — APPROVED** [review]
thanks, would be great to have the pickle example as a unit test, but I think it's beyond this PR.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

I found an easy way to test the pickle issue, already added the logic in the unit test in the latest commit.

Thanks.
